File: ACM/ACMProgrammingContest/武汉工程大学第一届程序设计女生赛/Code/B1.cpp
# include <iostream>
# include <set>

template <typename T>
void read(T &val) {
    T x = 0;
    int bz = 1;
    char c;
     
    for (c = getchar(); (c<'0' || c>'9') && c != '-';
    c = getchar());
     
    if (c == '-') {
        bz = -1;
        c = getchar();
    }
     
    for (; c >= '0' && c <= '9'; c = getchar())
        x = x * 10 + c - 48;
         
    val = x * bz;
}
 
const int MAXN = 100000 + 5;
int f[MAXN];
int s[MAXN];
 
int findfa(int x){
    return (f[x] == x) ? x : f[x] = findfa(f[x]);
}
 
void merge(int x, int y) {
    f[x] = y;
}
 
int main(){
    int T;
    std::cin >> T;
     
    while (T--) {
        int n, m, t;
        read(n);
        read(m);
        read(t);
        int cnt = 0;
         
        for(int i=1;i<=n;i++){
            f[i] = i;
        }
         
        for (int i=1; i<=m; i++) {
            read(s[i]);
        }
         
        for(int i=1; i<=t; i++){
            int u, v;
            read(u);
            read(v);
             
            if(findfa(u) != findfa(v)){
                merge(f[u], f[v]);
            }
        }
         
        std::set<int> st;
        for (int i=1; i<=m; i++) {
            st.insert(findfa(s[i]));
        }
         
        for (int i=1; i<=n; i++) {
            if (st.find(findfa(f[i])) != st.end()) {
                ++cnt;
            }
        }
         
        std::cout << cnt << "\n";
    }
    
    return 0;
}

File: ACM/ACMProgrammingContest/武汉工程大学第一届程序设计女生赛/Code/B1.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static final int MAXN = 100000 + 5;
    static int [] f = new int [MAXN];
    static int [] s = new int [MAXN];
     
    static int findfa(int x) {
        if (f[x] == x) {
            return x;
        }
         
        return f[x] = findfa(f[x]);
    }
     
    static void merge(int x, int y) {
        f[x] = y;
    }
     
    public static void main(String[] args) throws IOException  {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
         
        String line = in.readLine();
        int T = Integer.valueOf(line);
         
        while (T-- > 0) {
            line = in.readLine();
            String[] str = line.split(" ");
            int n = Integer.valueOf(str[0]);
            int m = Integer.valueOf(str[1]);
            int t = Integer.valueOf(str[2]);
            int cnt = 0;
             
            for (int i=1; i<=n; i++) {
                f[i] = i;
            }
             
            line = in.readLine();
            str = line.split(" ");
            for (int i=1; i<=m; i++) {
                s[i] = Integer.valueOf(str[i-1]);
            }
             
            for (int i=1; i<=t; i++) {          
                line = in.readLine();
                str = line.split(" ");
                int u = Integer.valueOf(str[0]);
                int v = Integer.valueOf(str[1]);
                 
                if (findfa(u) != findfa(v)) {
                    merge(f[u], f[v]);
                }
            }
             
            Set<Integer> st = new TreeSet<> ();
             
            for (int i=1; i<=m; i++) {
                st.add(findfa(s[i]));
            }
             
            for (int i=1; i<=n; i++) {
                if (st.contains(findfa(f[i]))) {
                    ++cnt;
                }
            }
             
            System.out.println(cnt);
        }
    }
}